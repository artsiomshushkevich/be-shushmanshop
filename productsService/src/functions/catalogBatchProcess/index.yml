catalogBatchProcess:
    handler: src/functions/catalogBatchProcess/catalogBatchProcess.main
    events:
        - sqs:
              arn:
                  Fn::GetAtt:
                      - ${self:custom.QUEUE_NAME}
                      - Arn
              batchSize: 5
              filterPatterns:
                  - body:
                        {
                            'title': [{ 'exists': true }],
                            'description': [{ 'exists': true }],
                            'price': [{ 'numeric': ['>', 0] }]
                        }
