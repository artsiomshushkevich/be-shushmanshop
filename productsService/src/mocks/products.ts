import type { Product } from "@localtypes/product";

export const products: Product[] = [
    {
        id: '1',
        title: 'Audi',
        description: 'E-tron 2021',
        price: 30000
    },
    {
        id: '2',
        title: 'Volkswagen',
        description: 'Golf E',
        price: 25000
    },
    {
        id: '3',
        title: 'BYD',
        description: 'Excellence',
        price: 45000
    },
    {
        id: '4',
        title: 'Tesla',
        description: 'Model S 2022',
        price: 57000
    }
];
