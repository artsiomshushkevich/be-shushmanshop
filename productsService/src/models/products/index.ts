export * from './product.model';
export * from './product.types';
export * from './products.mySqlModel';
