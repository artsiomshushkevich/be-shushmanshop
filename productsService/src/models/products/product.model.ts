import { Product } from './product.types';
import { db, unmarshall, ScanCommand, GetItemCommand } from '@libs/dynamodb';

export const productsModel = {
    getList: async (): Promise<Product[]> => {
        const response = await db.send(new ScanCommand({
            TableName:  'ProductsTable'
        }));

        return response.Items.map(item => unmarshall(item) as Product);
    },
    getById: async (id: string): Promise<Product | null> => {
        const response = await db.send(new GetItemCommand({
            TableName:  'ProductsTable',
            Key: { id: { S: id } }
        }));

        return response.Item ? unmarshall(response.Item) as Product : null;
    }
};
