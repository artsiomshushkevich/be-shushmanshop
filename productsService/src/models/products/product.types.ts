export type Product = {
    id: string;
    title: string;
    description: string;
    price: number;
};

export type ProductWithCount = Product & { count: number };

export type ProductWithCountWithoutId = Omit<ProductWithCount, 'id'>;
