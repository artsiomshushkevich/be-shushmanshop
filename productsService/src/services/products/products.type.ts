import type { Product } from '@models/products';

export type ProductWithCount = Product & { count: number };
