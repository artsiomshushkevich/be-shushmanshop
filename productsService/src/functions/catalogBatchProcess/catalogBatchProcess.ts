import { v4 as uuidv4 } from 'uuid';
import type { SQSHandler } from 'aws-lambda';
import { PublishCommand } from '@aws-sdk/client-sns';
import { DeleteMessageBatchCommand } from '@aws-sdk/client-sqs';
import { productsService } from '@services/products';
import { ProductWithCountWithoutId } from '@models/products';
import { sqsClient } from '@libs/sqs';
import { snsClient } from '@libs/sns';

export const main: SQSHandler = async (event) => {
    try {
        console.log('Getting all products from the queue. Records length', JSON.stringify(event.Records));
        const productsWithAmount: ProductWithCountWithoutId[] = event.Records.map((item) => {
            const unformattedProduct = JSON.parse(item.body);

            return {
                title: unformattedProduct.title,
                description: unformattedProduct.description,
                price: +unformattedProduct.price,
                count: +unformattedProduct.count || 0
            };
        });

        console.log('Saving all the products to DB', JSON.stringify(productsWithAmount));
        await productsService.bulkCreate(productsWithAmount);

        console.log('Deleting messages from queue');

        await sqsClient.send(
            new DeleteMessageBatchCommand({
                QueueUrl: process.env.QUEUE_URL,
                Entries: event.Records.map((item) => ({ Id: uuidv4(), ReceiptHandle: item.receiptHandle }))
            })
        );

        console.log('Sending notification to creators email');

        await snsClient.send(
            new PublishCommand({
                TopicArn: process.env.CREATE_PRODUCT_TOPIC_ARN,
                Message: `Products have been loaded! Date: ${new Date()}`
            })
        );
    } catch (e) {
        console.error('Failed to process products from SQS queue! ', e);
    }
};
