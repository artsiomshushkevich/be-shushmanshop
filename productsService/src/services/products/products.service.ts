import { v4 as uuidv4 } from 'uuid';
import { Product, productsModel, ProductWithCount, ProductWithCountWithoutId } from '@models/products';
import { Stock, stocksModel } from '@models/stocks';

export const productsService = {
    getList: async (): Promise<ProductWithCount[]> => {
        const result = await productsModel.getList();
        return result;
    },
    getById: async (id: string): Promise<ProductWithCount | null> => {
        const result = await productsModel.getById(id);
        return result;
    },
    create: async (product: ProductWithCountWithoutId): Promise<ProductWithCount> => {
        const id: string = uuidv4();
        const { title, description, price, count } = product;

        const newProduct = {
            id,
            title,
            description,
            price
        };

        const newProductWithCount = { ...newProduct, count: count || 0 };

        const newStock = {
            productId: id,
            count: count || 0
        };

        await productsModel.create(newProduct);
        await stocksModel.create(newStock);

        return newProductWithCount;
    },
    bulkCreate: async (products: Array<ProductWithCountWithoutId>): Promise<ProductWithCount[]> => {
        const productsToCreate: Product[] = [];
        const stocksToCreate: Stock[] = [];
        const productsWithCount: ProductWithCount[] = [];

        products.forEach((item) => {
            const id: string = uuidv4();
            const { title, description, price, count } = item;
            const formattedCount = count || 0;

            const newProduct: Product = {
                id,
                title,
                description,
                price
            };

            const newProductWithCount: ProductWithCount = { ...newProduct, count: formattedCount };

            const newStock: Stock = {
                productId: id,
                count: formattedCount
            };

            productsToCreate.push(newProduct);
            stocksToCreate.push(newStock);
            productsWithCount.push(newProductWithCount);
        });

        await productsModel.bulkCreate(productsToCreate);
        await stocksModel.bulkCreate(stocksToCreate);

        return productsWithCount;
    }
};
