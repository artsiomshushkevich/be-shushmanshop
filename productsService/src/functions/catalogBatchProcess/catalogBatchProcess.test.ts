import { main } from './catalogBatchProcess';
import { productsService } from '@services/products';

jest.mock('uuid', () => ({ v4: () => '916eb106-2ecc-4ac9-af2c-044f75220c5b' }));
jest.mock('@libs/sns', () => {
    return {
        snsClient: {
            send: () => Promise.resolve(true)
        }
    };
});

jest.mock('@libs/sqs', () => {
    return {
        sqsClient: {
            send: () => Promise.resolve(true)
        }
    };
});

const records = {
    Records: [
        {
            body: '{"title":"BYD","description":"Excellence","price":3000,"count":20}',
            receiptHandle: 'recipieptHandle0'
        },
        {
            body: '{"title":"Testla","description":"Model S","price":30000,"count":20}',
            receiptHandle: 'recipieptHandle1'
        }
    ]
};

describe('catalogBatchProcess', () => {
    let originEnv = process.env;

    beforeAll(() => {
        process.env = {
            ...originEnv,
            QUEUE_URL: 'https://aws.queue.url.com',
            CREATE_PRODUCT_TOPIC_ARN: 'arn:aws:XXXXX'
        };
    });

    beforeEach(() => {
        console.log = jest.fn().mockImplementation();
        console.error = jest.fn().mockImplementation();
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    test('Should successfully perform all the operations (getting products from records/creating in DB, deleting from SNS, sending SNS) and send logs', async () => {
        productsService.bulkCreate = jest.fn().mockReturnValue(Promise.resolve(true));
        // @ts-ignore
        await main(records);

        // @ts-ignore
        expect(console.log.mock.calls).toMatchSnapshot();
        // @ts-ignore
        expect(console.error.mock.calls).toMatchSnapshot();
    });

    test("Should make some logs until DB bulk creation won't thrown an error", async () => {
        productsService.bulkCreate = jest.fn().mockReturnValue(Promise.reject(false));
        // @ts-ignore
        await main(records);

        // @ts-ignore
        expect(console.log.mock.calls).toMatchSnapshot();

        // @ts-ignore
        expect(console.error.mock.calls).toMatchSnapshot();
    });
});
