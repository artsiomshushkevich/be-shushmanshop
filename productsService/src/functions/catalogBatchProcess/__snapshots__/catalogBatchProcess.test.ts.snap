// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalogBatchProcess Should make some logs until DB bulk creation won't thrown an error 1`] = `
[
  [
    "Getting all products from the queue. Records length",
    "[{"body":"{\\"title\\":\\"BYD\\",\\"description\\":\\"Excellence\\",\\"price\\":3000,\\"count\\":20}","receiptHandle":"recipieptHandle0"},{"body":"{\\"title\\":\\"Testla\\",\\"description\\":\\"Model S\\",\\"price\\":30000,\\"count\\":20}","receiptHandle":"recipieptHandle1"}]",
  ],
  [
    "Saving all the products to DB",
    "[{"title":"BYD","description":"Excellence","price":3000,"count":20},{"title":"Testla","description":"Model S","price":30000,"count":20}]",
  ],
]
`;

exports[`catalogBatchProcess Should make some logs until DB bulk creation won't thrown an error 2`] = `
[
  [
    "Failed to process products from SQS queue! ",
    false,
  ],
]
`;

exports[`catalogBatchProcess Should successfully perform all the operations (getting products from records/creating in DB, deleting from SNS, sending SNS) and send logs 1`] = `
[
  [
    "Getting all products from the queue. Records length",
    "[{"body":"{\\"title\\":\\"BYD\\",\\"description\\":\\"Excellence\\",\\"price\\":3000,\\"count\\":20}","receiptHandle":"recipieptHandle0"},{"body":"{\\"title\\":\\"Testla\\",\\"description\\":\\"Model S\\",\\"price\\":30000,\\"count\\":20}","receiptHandle":"recipieptHandle1"}]",
  ],
  [
    "Saving all the products to DB",
    "[{"title":"BYD","description":"Excellence","price":3000,"count":20},{"title":"Testla","description":"Model S","price":30000,"count":20}]",
  ],
  [
    "Deleting messages from queue",
  ],
  [
    "Sending notification to creators email",
  ],
]
`;

exports[`catalogBatchProcess Should successfully perform all the operations (getting products from records/creating in DB, deleting from SNS, sending SNS) and send logs 2`] = `[]`;
